File: tests/performance/drivers/python/app/results.py
"""Results output and CSV formatting."""

import csv
import json
import time
from pathlib import Path


def write_csv_results(results, test_name, driver_type):
    """Write test results to CSV file.
    
    Args:
        results: List of result dictionaries
        test_name: Name of the test
        driver_type: Driver type (universal or old)
    
    Returns:
        Path: Path to the created CSV file
    """
    timestamp = int(time.time())
    results_dir = Path("/results")
    results_dir.mkdir(exist_ok=True)
    
    filename = results_dir / f"{test_name}_python_{driver_type}_{timestamp}.csv"
    
    with open(filename, 'w', newline='') as f:
        writer = csv.DictWriter(f, fieldnames=["query_time_s", "fetch_time_s"])
        writer.writeheader()
        for result in results:
            writer.writerow({
                "query_time_s": f"{result['query_time_s']:.6f}",
                "fetch_time_s": f"{result['fetch_time_s']:.6f}",
            })
    
    return filename


def get_architecture():
    """Detect architecture inside the container."""
    import platform
    
    machine = platform.machine().lower()
    
    if machine in ('x86_64', 'amd64', 'x64'):
        return 'x86_64'
    elif machine in ('arm64', 'aarch64', 'armv8'):
        return 'arm64'
    else:
        return machine


def get_os_version():
    """Get OS version from environment variable (exported at container startup)."""
    import os
    return os.environ.get('OS_INFO', 'Linux')


def write_run_metadata(driver_type, driver_version, server_version):
    """Write run metadata JSON file (once per run).
    
    Args:
        driver_type: Driver type (universal or old)
        driver_version: Version string of the driver
        server_version: Snowflake server version
    """
    results_dir = Path("/results")
    metadata_filename = results_dir / f"run_metadata_python_{driver_type}.json"
    
    # Only write if doesn't exist (shared across all tests in run)
    if metadata_filename.exists():
        return
    
    # Detect architecture and OS inside container
    architecture = get_architecture()
    os_info = get_os_version()
    
    timestamp = int(time.time())
    metadata = {
        "driver": "python",
        "driver_type": driver_type,
        "driver_version": driver_version,
        "server_version": server_version,
        "architecture": architecture,
        "os": os_info,
        "run_timestamp": timestamp,
    }
    
    with open(metadata_filename, 'w') as f:
        json.dump(metadata, f, indent=2)

File: tests/performance/drivers/nodejs/app/results.js
'use strict';

const fs = require('fs');
const path = require('path');

function _resultsDir(driverType, testName) {
  const subdir = testName.endsWith('_record') ? '_record' : testName;
  const dir = path.join('/results', driverType, subdir);
  fs.mkdirSync(dir, { recursive: true });
  return dir;
}

function writeCsvResults(results, testName, driverType) {
  const timestamp = Math.floor(Date.now() / 1000);
  const dir = _resultsDir(driverType, testName);
  const filename = path.join(dir, `${testName}_nodejs_${driverType}_${timestamp}.csv`);

  const header = 'timestamp_ms,fetch_s,row_count,cpu_time_s,peak_rss_mb\n';
  const rows = results.map(
    (r) => `${r.timestamp},${r.fetchTimeS.toFixed(6)},${r.rowCount},${r.cpuTimeS.toFixed(6)},${r.peakRssMb.toFixed(1)}`,
  );
  fs.writeFileSync(filename, header + rows.join('\n') + '\n');
  return filename;
}

function writeColdStartResults(rows, testName, driverType) {
  const timestamp = Math.floor(Date.now() / 1000);
  const dir = _resultsDir(driverType, testName);
  const filename = path.join(dir, `${testName}_nodejs_${driverType}_${timestamp}.csv`);

  const header = 'timestamp_ms,e2e_s,load_s,connect_s,select1_s,cpu_time_s,peak_rss_mb\n';
  fs.writeFileSync(filename, header + rows.join('\n') + '\n');
  return filename;
}

function writeMemoryTimeline(memoryTimeline, testName, driverType) {
  if (!memoryTimeline || !memoryTimeline.length) return null;

  const timestamp = Math.floor(Date.now() / 1000);
  const dir = _resultsDir(driverType, testName);
  const filename = path.join(dir, `memory_timeline_${testName}_nodejs_${driverType}_${timestamp}.csv`);

  const header = 'timestamp_ms,rss_bytes,vm_bytes\n';
  const rows = memoryTimeline.map((s) => `${s.timestampMs},${s.rssBytes},${s.vmBytes}`);
  fs.writeFileSync(filename, header + rows.join('\n') + '\n');
  return filename;
}

function writeRunMetadata(driverType, driverVersion, serverVersion) {
  const metadataFilename = path.join('/results', `run_metadata_nodejs_${driverType}.json`);
  if (fs.existsSync(metadataFilename)) return;

  const metadata = {
    driver: 'nodejs',
    driver_type: driverType,
    driver_version: driverVersion,
    runtime_language_version: process.version,
    server_version: serverVersion,
    architecture: _getArchitecture(),
    os: process.env.OS_INFO || 'Linux',
    run_timestamp: Math.floor(Date.now() / 1000),
  };

  if (driverType === 'universal') {
    metadata.build_rust_version = process.env.BUILD_RUST_VERSION || 'NA';
  }

  fs.writeFileSync(metadataFilename, JSON.stringify(metadata, null, 2));
}

function _getArchitecture() {
  const arch = process.arch; // 'x64', 'arm64', etc.
  if (arch === 'x64') return 'x86_64';
  if (arch === 'arm64') return 'arm64';
  return arch;
}

module.exports = { writeCsvResults, writeColdStartResults, writeMemoryTimeline, writeRunMetadata };
